### PR TITLE
[SG-69] Fix type checking not working for Button component

### PR DESCRIPTION
**Pull request by igor-p**
This fixes the issue of the new `Button` component not recognizing any type signature for its props, which led to not having auto-complete/type-checking when entering props.

Part of the complexity was having to use `forwardRef` to pass-through the `ref` and needing to take a `component` prop of indeterminate type (which mandates that the Button be a generic component) in order to render the button as a custom component (e.g. a link). `forwardRef` returns a non-generic component, which prevented me from passing a `component` prop and allowing me to use props from that component type (e.g. `to` when using `RouterLink`) , and it messed with some other things, so I had to hunt around for a solution.

Props to the answer in https://github.com/mui-org/material-ui/issues/15827#issuecomment-809209533 for something that ultimately worked.

### CODE REVIEW

**Review by ironprice91 — DISMISSED** [review]
Nice, this is awesome! Can you update the package.json as well?

File: src/components/button/index.tsx [hunk 1]
```diff
@@ -1,91 +1,119 @@
...
-    { children, className, color, size = 'medium', variant, margin, ...muiProps }: ButtonProps<C>,
-    ref: React.Ref<HTMLButtonElement>
-  ) => {
-    const classes: StyleClasses = useStyles();
-    const classNames: string[] = [classes.root];
+type ModifiedTypeMap<P = {}, D extends React.ElementType = 'button'> = ExtendButtonBaseTypeMap<
+  Modify<
+    MuiButtonTypeMap<P, D>,
+    {
+      props: Modify<
```
Comment (herrethan) [inline src/components/button/index.tsx:17]: 👀 ! nice!

**Review by herrethan — APPROVED** [review]

File: src/components/button/index.tsx [hunk 2]
```diff
@@ -1,91 +1,119 @@
...
+        MuiButtonTypeMap<P, D>['props'],
+        {
+          color?: MuiButtonTypeMap['props']['color'] | 'tertiary' | 'success';
+          margin?: number | number[];
+          size?: MuiButtonTypeMap['props']['size'] | 'huge';
+          tabIndex?: number;
```
Comment (herrethan) [inline src/components/button/index.tsx:23]: Do we need this, is this just replicating MUI's props?

Comment (igor-p) [reply src/components/button/index.tsx]: For some reason, the compiler had a fit if I didn't include `tabIndex` here as a number. I just tried it now and it works fine with it removed. Must have been a relic of one of my previous attempts.

**Review by ironprice91 — APPROVED** [review]

**Review by herrethan — APPROVED** [review]

**Pull request merged**